src/GrammarType.hs

```diff
@@ -6,12 +6,16 @@ module GrammarType where
 
 import Data.Set (Set)
 
+-- |'Operand' is a
+data Operand = Conjunction | Negation
+	deriving (Eq, Ord)
+
 -- |'Terminal' is a type that represents terminal in the formal grammar 'Grammar'.
 newtype Terminal = Terminal String
     deriving (Eq, Ord)
 
 -- |'Nonterminal' is a type that represents nonterminal in the formal grammar 'Grammar'.
-newtype Nonterminal = Nonterminal String
+data Nonterminal = Nonterminal String | BooleanNonterminal String Operand
```

Comment:
What is booleanNonterminal? Nonterminal is a classical part of the grammar definition, booleanNonterminal looks strange.

Reply:
Ah.... I see. Try to realize, that conjunction and negation in grammar are the same as in boolean formulas, thus the structure of AST should be pretty similar.